### PR TITLE
fixes #12747 [backport]

### LINKED ISSUE #12747: Arrays are passed by copy to iterators, causing crashes, unnecessary allocations and slowdowns
```nim
type
    ABC = ref object
        arr: array[0x40000, pointer]

let a = ABC()

for a in a.arr:
    assert a == nil
```
This code crashes with an unspecified error:
```
Error: execution of an external program failed: 'c:\Users\Kemal\Desktop\workframe\arg.exe '
```

If I interpret the C code correctly, the reason for this that the whole array is copied onto the stack, which can't hold the 2mb object.

### PR DESCRIPTION
**Pull request by cooldome**

### CODE REVIEW

File: compiler/transf.nim [hunk 1]
```diff
@@ -563,7 +563,7 @@ proc putArgInto(arg: PNode, formal: PType): TPutArgInto =
       return paDirectMapping    # XXX really correct?
                                 # what if ``arg`` has side-effects?
   case arg.kind
-  of nkEmpty..nkNilLit:
+  of nkEmpty..nkNilLit, nkDotExpr:
```
Comment (Araq) [inline compiler/transf.nim]: What if it's ``sideEffectHere().field``? Then we cannot use a direct mapping, can we?

Comment (cooldome) [reply compiler/transf.nim]: Yep, fixed

**Comment by cooldome** [comment]
I have also added nkObjConstr handling as well and also noticed that nkExprColonExpr is not handled for nkPar, nkTupleConstr.

**Comment by Araq** [comment]
Please rebase

**Comment by cooldome** [comment]
Ready for review

**Review by Clyybber — APPROVED** [review]

**Pull request merged**